recursion/core/src/poseidon2_wide/trace.rs

```diff
@@ -34,43 +39,71 @@ impl<F: PrimeField32, const DEGREE: usize> MachineAir<F> for Poseidon2WideChip<D
+            }
+        }
+        nb_rows += input.poseidon2_compress_events.len() * 2;
+
+        let nb_padded_rows = if self.pad {
+            next_power_of_two(nb_rows, self.fixed_log2_rows)
+        } else {
+            nb_rows
+        };
 
         let num_columns = <Poseidon2WideChip<DEGREE> as BaseAir<F>>::width(self);
+        let mut rows = vec![F::zero(); nb_padded_rows * num_columns];
 
-        // Populate the hash events.
+        // Populate the hash events.  We do this serially, since each absorb event could populate a different
+        // number of rows.  Also, most of the rows are populated by the compress events.
+        let mut row_cursor = 0;
         for event in &input.poseidon2_hash_events {
             match event {
                 Poseidon2HashEvent::Absorb(absorb_event) => {
-                    rows.extend(self.populate_absorb_event(absorb_event, num_columns, output));
+                    let num_absorb_elements = absorb_event.iterations.len() * num_columns;
+                    let absorb_rows = &mut rows[row_cursor..row_cursor + num_absorb_elements];
+                    self.populate_absorb_event(absorb_rows, absorb_event, num_columns, output);
+                    row_cursor += num_absorb_elements;
                 }
 
                 Poseidon2HashEvent::Finalize(finalize_event) => {
-                    rows.push(self.populate_finalize_event(finalize_event, num_columns));
+                    let finalize_row = &mut rows[row_cursor..row_cursor + num_columns];
+                    self.populate_finalize_event(finalize_row, finalize_event);
+                    row_cursor += num_columns;
                 }
             }
         }
 
         // Populate the compress events.
-        for event in &input.poseidon2_compress_events {
-            rows.extend(self.populate_compress_event(event, num_columns));
-        }
+        let compress_rows = &mut rows[row_cursor..];
+        par_for_each_row(compress_rows, 2, num_columns, |i, rows| {
+            if i >= input.poseidon2_compress_events.len() {
+                return;
+            }
+
+            self.populate_compress_event(rows, &input.poseidon2_compress_events[i], num_columns);
+        });
+
+        // Convert the trace to a row major matrix.
+        let mut trace = RowMajorMatrix::new(rows, num_columns);
+
+        let padded_rows = trace.values.par_chunks_mut(num_columns).skip(nb_rows);
 
         if self.pad {
-            // Pad the trace to a power of two.
-            pad_rows_fixed(
-                &mut rows,
-                || {
-                    let mut padded_row = vec![F::zero(); num_columns];
-                    self.populate_permutation([F::zero(); WIDTH], None, &mut padded_row);
-                    padded_row
-                },
-                self.fixed_log2_rows,
-            );
+            padded_rows.for_each(|padded_row| {
+                self.populate_permutation([F::zero(); WIDTH], None, padded_row);
```

Comment:
this could be slow if populate_permutation is slow? could be better to populate once and copy to all the rows